File: countVowelStrings/countVowelStrings.js
/**
 * @param {number} n
 * @return {number}
 */

// var hash = {}
// var countVowelStrings = function (n, currentNum=5) {
//   if(n == 1) return currentNum

//   let count = 0;
//   for(let i=5; i>=1; i--) {
//     count += countVowelStrings(n-1, i)
//   }

//   return count
// };

var countVowelStrings = function (n) {
  var dp = {}
  for (let i = 1; i <= n; i++) {
    dp[i] = {}
    for (let k = 1; k <= 5; k++) {
      dp[i][k] = ( k > 1 ? dp[i][k-1] : 0) + ( i > 1 ? dp[i-1][k] : 1)
    }
  }

  return dp[n][5];
}

console.log(countVowelStrings(33))

File: countVowelStrings/countVowelStrings.cpp
#include <iostream>
#include <vector>
using namespace std;
int main() {
  
vector<vector<int>> vec { { 1, 2, 3 }, 
                         { 4, 5, 6 }, 
                         { 7, 8, 9, 4 } }; 
  
}
// https://leetcode.com/problems/count-sorted-vowel-strings/
#include <iostream>
#include <vector>
using namespace std;
class Solution {
  public:
    int countVowelStrings(int n)
    {
      vector<vector<int> > dp(n + 1, vector<int>(6));
      for (int i = 1; i <= n; ++i)
        for (int k = 1; k <= 5; ++k)
          dp[i][k] = dp[i][k - 1] + (i > 1 ? dp[i - 1][k] : 1);
      return dp[n][5];
    }
};